core/src/banking_stage/latest_unprocessed_votes.rs

```diff
@@ -322,17 +330,30 @@ impl LatestUnprocessedVotes {
     }
 
     /// Drains all votes yet to be processed sorted by a weighted random ordering by stake
+    /// Do not touch votes that are for a different fork from `bank` as we know they will fail,
+    /// however the next bank could be built on a different fork and consume these votes.
```

Comment:
previously this method would clear out all the votes, right?

How do we get the unprocessed, but on a different fork, votes out of this container?

Reply:
If we later build a bank on a different fork, we will drain votes that were skipped over that now qualify.
Otherwise these votes will be replaced by newer votes coming in from gossip / tpu, we still only store one vote per Pubkey.

Reply:
Do we ever clear votes if we are not leader and not leader soon?

Reply:
Nope, as it is impossible to tell if that vote could be included in our next leader bank. 
The buffer for gossip votes (which I will remove next) attempts to do this by only holding votes from `3 * 512` slots before our leader, but even that is a probabilistic measure:
https://github.com/anza-xyz/agave/blob/784352c3765792105764c875c1a619e6f71ded55/core/src/cluster_info_vote_listener.rs#L395

Since the leader can choose to build off any block, we can place no lower bound on the block height of the fork. So a valid vote for this fork could have been sent any number of slots ago.